determinism.md

```diff
@@ -0,0 +1,174 @@
+GetDeterministicKey
+===
+
+**This is meant to be the text of the issue I will eventually file for the API**
+
+## Background and Motivation
+The `Compilation` type is fully deterministic meaning that given the same inputs
+(`SyntaxTree`, `CompilationOptions`, etc ...) it will produce the same output. By 
+that it means it will produce byte for byte equivalent binaries and the same 
+diagnostics. This is extremely valuable in infrastructure because it allows 
+for caching and opens the door for efficient distributed processing.
+
+At the moment this is hard to leverage because consumers can't determine if 
+two `Compilation` instances are equivalent for the purposes of determinism. 
+Customers have to resort to hand written comparisons which requires a fairly 
+intimate knowledge of the compiler (for example knowing what does and does not
+impact determinism). Such solutions are not version tolerant; every time the 
+compiler adds a new property that impacts determinism the solution must be 
+updated. Even when proper equality checks are in place this does not help 
+distributed computing where equality must be decided across different processes.
+
+The motivation here is to provide an API that returns a string based key for a
+given `Compilation` such that for two equivalent `Compilation` instances the
+key will also be equivalent. This will allow customers who want to leverage 
+deterministic caching and communication to have a simple, and transmittable,
+way to full describe the `Compilation` they are processing.
+
+## Proposed API
+
+```diff
+namespace Microsoft.CodeAnalysis
+{
+    public class Compilation
+    {
++       public static string GetDeterministicKey(
++           CompilationOptions compilationOptions,
++           ImmutableArray<SyntaxTree> syntaxTrees,
```

Comment:
definitely will talk about this in our meeting today (specifically teh SyntaxTree part).

Reply:
let's make sure to record what suggestion we have here. e.g. are we going to simply make the parameter `ImmutableArray<SourceText>` instead, and make you pass ParseOptions separately.

Reply:
Unfortunately, the compiler api allows each tree to have different parseoptions.  so i think we need some object that encapsulates the three things that are actually used here).

Reply:
Moving the API discussion to the issue now, made sure this was recorded in the new API on the issue